modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -48,37 +52,51 @@
-  (^Iterable [{{:keys [project-id dataset-id]} :details, :as database}]
-    (let [bq (database->client database)]
-      (list-tables bq project-id dataset-id)))
-
-  (^Iterable [^BigQuery client, ^String project-id ^String dataset-id]
-    {:pre [client (not (str/blank? dataset-id))]}
-    (.iterateAll (.listTables client (DatasetId/of project-id dataset-id) (u/varargs BigQuery$TableListOption)))))
+  (^Iterable [{{:keys [project-id dataset-filters-type dataset-filters-patterns]} :details, :as database}]
+    (list-tables (database->client database)
+                 (or project-id (bigquery.common/database-details->credential-project-id (:details database)))
+                 dataset-filters-type
+                 dataset-filters-patterns))
+
+  (^Iterable [^BigQuery client ^String project-id ^String filter-type ^String filter-patterns]
+    (let [datasets (.listDatasets client project-id (u/varargs BigQuery$DatasetListOption))
+          inclusion-patterns (when (= "inclusion" filter-type) filter-patterns)
+          exclusion-patterns (when (= "exclusion" filter-type) filter-patterns)]
+      (apply concat (for [^Dataset dataset (.iterateAll datasets)
+                          :let [^DatasetId dataset-id (.. dataset getDatasetId)]
+                          :when (describe-database/include-schema? (.getDataset dataset-id)
+                                                                   inclusion-patterns
+                                                                   exclusion-patterns)]
+                      (-> (.listTables client dataset-id (u/varargs BigQuery$TableListOption))
+                          .iterateAll
+                          .iterator
+                          iterator-seq))))))
 
 (defmethod driver/describe-database :bigquery-cloud-sdk
   [_ database]
-  (let [tables (list-tables database)]
-    {:tables (set (for [^Table table tables]
-                    {:schema nil, :name (.. table getTableId getTable)}))}))
+  (let [tables             (list-tables database)]
+    {:tables (set (for [^Table table tables
+                        :let  [^TableId table-id  (.getTableId table)
+                               ^String dataset-id (.getDataset table-id)]]
+                    {:schema dataset-id, :name (.getTable table-id)}))}))
 
 (defmethod driver/can-connect? :bigquery-cloud-sdk
-  [_ {:keys [project-id dataset-id] :as details-map}]
-  ;; check whether we can connect by seeing whether we have at least one Table in the iterator
-  (let [^BigQuery bq     (database->client {:details details-map})
-        ^DatasetId ds-id (if (some? project-id)
-                           (DatasetId/of project-id dataset-id)
-                           (DatasetId/of dataset-id))]
-    (some? (.getDataset bq ds-id (u/varargs BigQuery$DatasetOption)))))
+  [_ details-map]
+  ;; check whether we can connect by seeing whether listing tables succeeds
+  (try (some? (list-tables {:details details-map}))
```

Comment:
`some?` on a lazy seq will always return true since a lazy seq is an object and not null. 
```clojure
((juxt some? identity) (filter string? (range 5)))
-> [true ()]
```

Reply:
Yeah, this is deliberate.  Have a look at `can-connect?-test` in the test namespace.  Basically, we want to confirm that the list table operation had some result (even an empty sequence).  If there is an error in listing tables (ex: due to an authentication or networking problem) then an exception will be thrown instead and `can-connect?` will return false.

Reply:
oh so `some?` is used like `(try (list-tables) true (catch ...))`